File: projects-review/agent-trust/packages/python-sdk/agent_trust/scoring.py
"""
AgentTrust scoring — pure Python port of the TypeScript scoring algorithm.
Includes EWMA time-decay (30-day half-life), same as the TS implementation.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Literal, Optional


@dataclass
class ScoreDimensions:
    completion_rate: float = 0.0
    response_time: float = 0.0
    reliability_score: float = 0.0
    consistency_score: float = 0.0


@dataclass
class TransactionRecord:
    provider_did: str
    protocol: Literal["x402", "mcp", "a2a", "other"]
    status: Literal["success", "failure", "disputed"]
    response_time_ms: float
    amount_usd: float = 0.0
    consumer_did: Optional[str] = None
    id: Optional[str] = None
    created_at: Optional[str] = None

    def __post_init__(self):
        if self.id is None:
            import uuid
            self.id = str(uuid.uuid4())
        if self.created_at is None:
            self.created_at = datetime.now(timezone.utc).isoformat()


@dataclass
class TrustScore:
    did: str
    overall_score: float
    dimensions: ScoreDimensions
    transaction_count: int
    computed_at: str
    confidence_tier: Literal["insufficient_data", "low", "medium", "high"]

    @property
    def grade(self) -> str:
        if self.overall_score >= 90: return "A"
        if self.overall_score >= 75: return "B"
        if self.overall_score >= 60: return "C"
        if self.overall_score >= 40: return "D"
        return "F"


WEIGHTS = dict(
    completion_rate=0.35,
    reliability_score=0.30,
    consistency_score=0.20,
    response_time=0.15,
)

MIN_FOR_LOW = 5
MIN_FOR_MEDIUM = 25
MIN_FOR_HIGH = 100

EWMA_HALF_LIFE_DAYS = 30
EWMA_PRIOR_N = 10
EWMA_PRIOR_SCORE = 70


def _normalise_response_time(avg_ms: float) -> float:
    FAST, SLOW = 500, 10_000
    if avg_ms <= FAST: return 100.0
    if avg_ms >= SLOW: return 0.0
    return 100 * (1 - (avg_ms - FAST) / (SLOW - FAST))


def _ewma_consistency(records: list[TransactionRecord]) -> float:
    """EWMA time-decay consistency score (30-day half-life)."""
    if not records:
        return float(EWMA_PRIOR_SCORE)

    half_life_ms = EWMA_HALF_LIFE_DAYS * 24 * 3600 * 1000
    decay = math.log(2) / half_life_ms
    now_ms = datetime.now(timezone.utc).timestamp() * 1000

    # Prior pseudo-transactions weighted at 30-day age
    prior_weight = EWMA_PRIOR_N * math.exp(-decay * half_life_ms)
    weighted_sum = prior_weight * EWMA_PRIOR_SCORE
    total_weight = prior_weight

    for r in records:
        try:
            ts = datetime.fromisoformat(r.created_at.replace("Z", "+00:00")).timestamp() * 1000
        except Exception:
            ts = now_ms
        age = max(0, now_ms - ts)
        w = math.exp(-decay * age)
        value = 100 if r.status == "success" else (0 if r.status == "disputed" else 30)
        weighted_sum += w * value
        total_weight += w

    return round(weighted_sum / total_weight, 2)


def compute_trust_score(did: str, records: list[TransactionRecord]) -> TrustScore:
    """
    Compute an AgentTrust score for the given DID based on its transaction records.

    This is a pure-Python port of the TypeScript implementation in:
    packages/core/src/scoring.ts
    """
    now = datetime.now(timezone.utc).isoformat()

    if not records:
        return TrustScore(
            did=did,
            overall_score=0,
            dimensions=ScoreDimensions(),
            transaction_count=0,
            computed_at=now,
            confidence_tier="insufficient_data",
        )

    n = len(records)
    success_count = sum(1 for r in records if r.status == "success")
    dispute_count = sum(1 for r in records if r.status == "disputed")
    avg_ms = sum(r.response_time_ms for r in records) / n

    dims = ScoreDimensions(
        completion_rate=round(success_count / n * 100, 2),
        response_time=round(_normalise_response_time(avg_ms), 2),
        reliability_score=round(max(0, 100 - (dispute_count / n) * 200), 2),
        consistency_score=_ewma_consistency(records),
    )

    overall = round(
        dims.completion_rate   * WEIGHTS["completion_rate"]
        + dims.reliability_score * WEIGHTS["reliability_score"]
        + dims.consistency_score * WEIGHTS["consistency_score"]
        + dims.response_time     * WEIGHTS["response_time"],
        2,
    )

    tier: Literal["insufficient_data", "low", "medium", "high"] = (
        "insufficient_data" if n < MIN_FOR_LOW
        else "low"    if n < MIN_FOR_MEDIUM
        else "medium" if n < MIN_FOR_HIGH
        else "high"
    )

    return TrustScore(
        did=did,
        overall_score=overall,
        dimensions=dims,
        transaction_count=n,
        computed_at=now,
        confidence_tier=tier,
    )

File: projects-review/agent-trust/packages/core/src/scoring.ts
import type {
  AgentTrustScore,
  ScoreDimensions,
  TransactionRecord,
  DID,
} from './types.js';
import { getEcosystemMeta } from './ecosystem.js';

/**
 * Scoring weights — must sum to 1.0.
 * These are intentionally documented and version-controlled so anyone
 * can audit exactly how the composite score is computed.
 */
const WEIGHTS = {
  completionRate: 0.35,
  reliabilityScore: 0.30,
  consistencyScore: 0.20,
  responseTime: 0.15,
} as const;

/** Minimum transactions before we report anything above "insufficient_data" */
const MIN_TRANSACTIONS_FOR_LOW = 5;
const MIN_TRANSACTIONS_FOR_MEDIUM = 25;
const MIN_TRANSACTIONS_FOR_HIGH = 100;

/**
 * Compute a normalised 0-100 response-time score.
 * Benchmark: p50 of observed x402 transactions ~800ms.
 * Anything under 500ms scores 100; anything over 10s scores 0.
 */
function normaliseResponseTime(avgMs: number): number {
  const FAST_THRESHOLD = 500;
  const SLOW_THRESHOLD = 10_000;
  if (avgMs <= FAST_THRESHOLD) return 100;
  if (avgMs >= SLOW_THRESHOLD) return 0;
  return Math.round(
    100 * (1 - (avgMs - FAST_THRESHOLD) / (SLOW_THRESHOLD - FAST_THRESHOLD))
  );
}

/**
 * EWMA (Exponentially Weighted Moving Average) time-decay scoring.
 *
 * Borrowed from PeerClaw's reputation model. More recent transactions
 * carry more weight, preventing an agent from coasting on old good behavior.
 *
 * Half-life: 30 days — a transaction from 30 days ago contributes half the
 * weight of a transaction from today.
 *
 * See: Issue #1 "Should trust scores decay over time?"
 */
function computeEwmaConsistency(records: TransactionRecord[]): number {
  if (records.length === 0) return 70; // Bayesian prior

  const HALF_LIFE_MS = 30 * 24 * 60 * 60 * 1000; // 30 days
  const DECAY_FACTOR = Math.LN2 / HALF_LIFE_MS;
  const now = Date.now();

  // Prior: 10 pseudo-transactions at score 70, dated 30 days ago (half weight)
  const PRIOR_N = 10;
  const PRIOR_SCORE = 70;
  const PRIOR_WEIGHT = PRIOR_N * Math.exp(-DECAY_FACTOR * HALF_LIFE_MS); // ~5

  let weightedSum = PRIOR_WEIGHT * PRIOR_SCORE;
  let totalWeight = PRIOR_WEIGHT;

  for (const r of records) {
    const age = now - new Date(r.createdAt).getTime();
    const weight = Math.exp(-DECAY_FACTOR * Math.max(0, age));
    const value = r.status === 'success' ? 100 : r.status === 'disputed' ? 0 : 30;
    weightedSum += weight * value;
    totalWeight += weight;
  }

  return Math.round(weightedSum / totalWeight);
}

/**
 * Legacy Bayesian consistency (kept for backward-compat testing).
 * @deprecated Use computeEwmaConsistency instead.
 */
function _computeConsistencyBayesian(records: TransactionRecord[]): number {
  const PRIOR_N = 10;
  const PRIOR_SCORE = 70;
  if (records.length === 0) return PRIOR_SCORE;
  const successValues: number[] = records.map(r => (r.status === 'success' ? 100 : 0));
  const observed: number = successValues.reduce((a: number, b: number) => a + b, 0) / successValues.length;
  const smoothed =
    (PRIOR_N * PRIOR_SCORE + records.length * observed) /
    (PRIOR_N + records.length);
  return Math.round(smoothed);
}

/**
 * Main scoring function.
 * Pure — given the same records + same current time, produces the same score.
 *
 * v0.2: Added EWMA time-decay for consistency score (30-day half-life).
 */
export function computeTrustScore(
  did: DID,
  records: TransactionRecord[]
): AgentTrustScore {
  const now = new Date().toISOString();

  if (records.length === 0) {
    return {
      did,
      overallScore: 0,
      dimensions: {
        completionRate: 0,
        responseTime: 0,
        reliabilityScore: 0,
        consistencyScore: 0,
      },
      transactionCount: 0,
      computedAt: now,
      confidenceTier: 'insufficient_data',
      _ecosystem: getEcosystemMeta(),
    };
  }

  const successCount = records.filter(r => r.status === 'success').length;
  const disputeCount = records.filter(r => r.status === 'disputed').length;
  const avgResponseMs =
    records.reduce((sum, r) => sum + r.responseTimeMs, 0) / records.length;

  const dimensions: ScoreDimensions = {
    completionRate: Math.round((successCount / records.length) * 100),
    responseTime: normaliseResponseTime(avgResponseMs),
    reliabilityScore: Math.round(
      Math.max(0, 100 - (disputeCount / records.length) * 200)
    ),
    // v0.2: EWMA replaces naive Bayesian smoothing
    consistencyScore: computeEwmaConsistency(records),
  };

  const overallScore = Math.round(
    dimensions.completionRate * WEIGHTS.completionRate +
      dimensions.reliabilityScore * WEIGHTS.reliabilityScore +
      dimensions.consistencyScore * WEIGHTS.consistencyScore +
      dimensions.responseTime * WEIGHTS.responseTime
  );

  const n = records.length;
  const confidenceTier =
    n < MIN_TRANSACTIONS_FOR_LOW
      ? 'insufficient_data'
      : n < MIN_TRANSACTIONS_FOR_MEDIUM
      ? 'low'
      : n < MIN_TRANSACTIONS_FOR_HIGH
      ? 'medium'
      : 'high';

  return {
    did,
    overallScore,
    dimensions,
    transactionCount: n,
    computedAt: now,
    confidenceTier,
    _ecosystem: getEcosystemMeta(),
  };
}

/**
 * Scoring weights exported for transparency.
 * The MCP server surfaces these so callers can understand the formula.
 */
export const SCORING_WEIGHTS = WEIGHTS;

/**
 * Export EWMA parameters for auditability.
 */
export const EWMA_PARAMS = {
  halfLifeDays: 30,
  description: 'Transactions from 30 days ago carry half the weight of today\'s transactions.',
  priorTransactions: 10,
  priorScore: 70,
} as const;
